### PR TITLE
Orders

### PR DESCRIPTION
**Pull request by TatsianaMaltseva**
image
ability to select check-in and check-out dates is added
image
image

### CODE REVIEW

File: admin/src/app/Components/auth.interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import {
+  HttpRequest,
+  HttpHandler,
+  HttpEvent,
+  HttpInterceptor,
+  HttpErrorResponse
+} from '@angular/common/http';
+import { Observable, throwError } from 'rxjs';
+import { catchError } from 'rxjs/operators';
+import { MatDialog } from '@angular/material/dialog';
```
Comment (bagatka) [inline admin/src/app/Components/auth.interceptor.ts:11]: Group by

File: admin/src/app/Components/auth.interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import { AuthenticationDialogComponent } from './authenticationDialog/authenticationDialog.component';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+  public constructor(
+    private readonly matDialog: MatDialog
+  ) {
+  }
+  public intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```
Comment (bagatka) [inline admin/src/app/Components/auth.interceptor.ts]: Empty line between methods

File: admin/src/app/Components/auth.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        catchError((response: HttpErrorResponse) => {
+          if (response.status === 401) {
+            this.login();
+          }
+          return throwError(response);
+        })
+      );
+  }
+
+  public login(): void {
```
Comment (bagatka) [inline admin/src/app/Components/auth.interceptor.ts]: Can you think about better name?

File: admin/src/app/Components/hotel-for-admin/hotel-for-admin.component.html [hunk 4]
```diff
@@ -57,8 +68,17 @@
                     <mat-icon>add_shopping_cart</mat-icon>facilities
                 </button>
             </mat-card-actions>
-            <app-images-for-admin [hotel]="hotel"></app-images-for-admin>
-            <app-rooms-for-admin [hotel]="hotel"></app-rooms-for-admin>
+            <app-images-for-admin *ngIf="isHotelExistInDataBase" [hotel]="hotel"></app-images-for-admin>
+            <app-rooms-for-admin *ngIf="isHotelExistInDataBase" [hotel]="hotel"></app-rooms-for-admin>
+            <mat-card-actions *ngIf="!isHotelExistInDataBase">
```
Comment (bagatka) [inline admin/src/app/Components/hotel-for-admin/hotel-for-admin.component.html]: *ngIf/else?

File: admin/src/app/Components/order/order.component.html [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    <span *ngIf="facility.price; else free">- US${{ facility.price }}</span>
+    <ng-template #free>- free</ng-template>
+</mat-checkbox>
+<mat-checkbox 
+    *ngFor="let facility of order.hotel.facilities"
+    class="facilities"
+    (change)="changeHotelFacilityStatus($event, facility)"
+>
+    {{ facility.name }} 
+    <span *ngIf="facility.price; else free">- US${{ facility.price }}</span>
```
Comment (bagatka) [inline admin/src/app/Components/order/order.component.html]:
```suggestion
    <span>- US${{ facility.price ?? 'free' }}</span>
```

File: admin/src/app/Components/order/order.component.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import { Component, Inject } from '@angular/core';
+import { MatCheckboxChange } from '@angular/material/checkbox';
+import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
+import { Facility } from 'src/app/Dtos/facility';
+import * as dayjs from 'dayjs';
+
+import { Order } from 'src/app/Dtos/order';
+import { OrderDetails } from 'src/app/Dtos/order-details';
+import { HotelFilterService } from 'src/app/hotel-filter.service';
+import { OrderService } from 'src/app/orders.service';
```
Comment (bagatka) [inline admin/src/app/Components/order/order.component.ts]: Group by

File: admin/src/app/Components/order/order.component.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      hotel: orderDetails.hotel,
+      orderDateParams: {
+        checkInDate: this.hotelFilterService.checkInDate,
+        checkOutDate: this.hotelFilterService.checkOutDate
+      },
+      price: orderDetails.room.price
+    };
+    this.days = dayjs(this.order.orderDateParams.checkOutDate).diff(
+      dayjs(this.order.orderDateParams.checkInDate), 
+      'day') + 1;
```
Comment (bagatka) [inline admin/src/app/Components/order/order.component.ts]: Code style

File: admin/src/app/Components/rooms/rooms.component.html [hunk 8]
```diff
@@ -1,64 +1,63 @@
...
-                        class="img__img" 
-                        [src]="createImagePath(room)"
-                    />
-                    <p class="img__description" (click)="showImagesDialog(room)">
-                        Click to see other images
-                    </p>
-                </div>
-            </td>
-        </ng-container>
+<p> Rooms available from {{ checkInDate }} to {{ checkOutDate }}</p>
```
Comment (bagatka) [inline admin/src/app/Components/rooms/rooms.component.html]:
```suggestion
<p>Rooms available from {{ checkInDate }} to {{ checkOutDate }}</p>
```

File: admin/src/app/Components/rooms/rooms.component.ts [hunk 9]
```diff
@@ -1,98 +1,114 @@
...
+        data: { room, hotel: this.hotel } as OrderDetails
+      }
+    );
+
+    dialogRef
+      .afterClosed()
+      .subscribe(() => {
+          this.rooms = this.rooms.filter(room => room.number > 0);
+        }
+    );
```
Comment (bagatka) [inline admin/src/app/Components/rooms/rooms.component.ts]:
```suggestion
      .subscribe(() => {
          this.rooms = this.rooms.filter(room => room.number > 0);
        }
      );
```

File: admin/src/app/hotel.service.ts [hunk 10]
```diff
@@ -38,18 +41,32 @@ export class HotelService {
...
-      { ...editedHotel }
+      editedHotel
     );
   }
 
   public getHotelsCount(filterParameters: HotelFilterParameters): Observable<number> {
     const httpParams = filterParameters as Params;
-    return this.http.get<number>(`${this.apiUrl}api/hotels/count`, { params: httpParams });
+    return this.http.get<number>(`${this.apiUrl}api/hotels/count`,
+    { params: httpParams });
```
Comment (bagatka) [inline admin/src/app/hotel.service.ts]: Code style

File: api/Controllers/OrdersController.cs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        [Authorize(Roles = Role.Client)]
+        public async Task<IActionResult> AddOrder([FromBody] OrderToAdd order)
+        {
+            if (!await CheckIfRoomExists(order.Room.Id))
+            {
+                return BadRequest("Such room does not exist");
+            }
+
+            var facilities = new List<FacilityEntity>();
+            foreach(Facility facility in order.Room.Facilities)
```
Comment (bagatka) [inline api/Controllers/OrdersController.cs:40]:
```suggestion
            foreach (Facility facility in order.Room.Facilities)
```

File: api/Controllers/OrdersController.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+            {
+                if (facility.Checked == true)
+                {
+                    facilities.Add(await _hotelsDb.Facilities.FirstOrDefaultAsync(facilityEntity => facilityEntity.Id == facility.Id));
+                }
+            }
+
+            foreach(Facility facility in order.Hotel.Facilities)
```
Comment (bagatka) [inline api/Controllers/OrdersController.cs:48]:
```suggestion
            foreach (Facility facility in order.Hotel.Facilities)
```

File: api/Controllers/OrdersController.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        public OrdersController(HotelsDatabaseContext hotelDb, IMapper mapper)
+        {
+            _hotelsDb = hotelDb;
+            _mapper = mapper;
+        }
+
+        [Route("orders")]
+        [HttpPost]
+        [Authorize(Roles = Role.Client)]
+        public async Task<IActionResult> AddOrder([FromBody] OrderToAdd order)
```
Comment (bagatka) [inline api/Controllers/OrdersController.cs:32]: Why do you need to send facilities that were not marked as selected

Comment (TatsianaMaltseva) [reply api/Controllers/OrdersController.cs:32]: It Is better to do in in frontend application, I am agree with you

**Pull request closed without merging**